clippy_lints/src/entry.rs

```diff
@@ -56,132 +56,597 @@ declare_clippy_lint! {
 declare_lint_pass!(HashMapPass => [MAP_ENTRY]);
 
 impl<'tcx> LateLintPass<'tcx> for HashMapPass {
+    #[allow(clippy::too_many_lines)]
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
-        if let ExprKind::If(ref check, ref then_block, ref else_block) = expr.kind {
-            if let ExprKind::Unary(UnOp::Not, ref check) = check.kind {
-                if let Some((ty, map, key)) = check_cond(cx, check) {
-                    // in case of `if !m.contains_key(&k) { m.insert(k, v); }`
-                    // we can give a better error message
-                    let sole_expr = {
-                        else_block.is_none()
-                            && if let ExprKind::Block(ref then_block, _) = then_block.kind {
-                                (then_block.expr.is_some() as usize) + then_block.stmts.len() == 1
-                            } else {
-                                true
-                            }
-                        // XXXManishearth we can also check for if/else blocks containing `None`.
-                    };
-
-                    let mut visitor = InsertVisitor {
-                        cx,
-                        span: expr.span,
-                        ty,
-                        map,
-                        key,
-                        sole_expr,
-                    };
-
-                    walk_expr(&mut visitor, &**then_block);
+        let (cond_expr, then_expr, else_expr) = match expr.kind {
+            ExprKind::If(c, t, e) => (c, t, e),
+            _ => return,
+        };
+        let (map_ty, contains_expr) = match try_parse_contains(cx, cond_expr) {
+            Some(x) => x,
+            None => return,
+        };
+
+        let mut app = Applicability::MachineApplicable;
+        let map_str = snippet_with_context(cx, contains_expr.map.span, contains_expr.call_ctxt, "..", &mut app).0;
+        let key_str = snippet_with_context(cx, contains_expr.key.span, contains_expr.call_ctxt, "..", &mut app).0;
+        let sugg = if let Some(else_expr) = else_expr {
```

Comment:
To help us quickly understand which branch corresponds to which suggestion, could you add some comments or split these into some methods which have an appropriate name?

Reply:
Added some comments and shortened it a little. Would a short overview of what's being checked for help here?

Reply:
@Jarcho Thanks for your adding comments!

I started the review again, but, I found that it would still take a long time for me to review because the fixes were large and there were multiple fixes / improvements together. 
So, if possible, could you split this PR into multiple pieces? I think it would be better to have one PR for one bug fix / improvement.

Reply:
All of those are actually fixed with the same change.

I could split out generating the `insert_with` closure. Most of the complexity of the visitor is due to that.

Reply:
Often times it also helps to split large PRs up into multiple commits. That way, the reviewer can review the command commit-by-commit, while not every commit has to actually build/pass tests.

Reply:
First pass at splitting it up is done. Let me know if you'd like it done more.

Reply:
Thanks for splitting commits!